File: benchmark/synduce/compressed_list/sum.f
Config NonLinear = true;

Inductive Nat = z Unit | s Nat;
Inductive List = nil Unit | cons {Int, List};
Inductive CList = cnil Unit | ccons {Nat, Int, CList};

repeat = \w: Int. \xs: List. fix (
  \f: Nat -> List. \n: Nat.
  match n with
    z _ -> xs
  | s m -> cons {w, f m}
  end
);

value = fix (
  \f: Nat -> Int. \n: Nat.
  match n with
    z _ -> 0
  | s m -> + 1 (f m)
  end
);

spec = fix (
  \f: List -> Int. \xs: List.
  match xs with
    nil _ -> 0
  | cons {h, t} -> + h (f t)
  end
);

repr = fix (
  \f: CList -> Compress List. \xs: CList.
  match xs with
    cnil _ -> nil unit
  | ccons {n, h, t} ->
    /* (value n) is given in Synduce's template*/
    let w = value n in
      repeat h (f t) n
  end
);

main = \xs: CList. spec (repr xs);

File: ml-benchmark/synduce/compressed_list/sum.ml
config NonLinear = true

type nat =
  | Z
  | S of nat

type list =
  | Nil
  | Cons of int * list

type compressed_list =
  | CNil
  | CCons of nat * int * compressed_list

let rec repeat a cont = function
  | Z -> cont
  | S n -> Cons (a, repeat a cont n)

let rec stutter = function
  | CNil -> Nil
  | CCons (hdc, hdv, tl) -> repeat hdv (stutter tl) hdc

let rec sum_list = function
  | Nil -> 0
  | Cons (hd, tl) -> hd + sum_list tl

let rec value = function
  | Z -> 0
  | S n -> 1 + value n

val repr: compressed_list -> list compress 
let rec repr = function 
  | CNil -> Nil
  | CCons (n, h, t) ->
    let w = value n in 
    repeat h (repr t) n 

let prog xs = sum_list (repr xs)
  